_episodes/38.md

```diff
@@ -0,0 +1,475 @@
+
+#### Pivot
+
+Pivot is an oft-requested feature which hasn't been built in Trino because it
+isn't a part of the standard SQL specification. A `PIVOT` keyword or built-in
+function isn't planned, but with PTFs, we can support `PIVOT`-like functionality
+without needing to deviate from SQL.
+
+A `PIVOT` PTF might have the following definition:
+
+```sql
+FUNCTION Pivot (
+    Input_table TABLE PASS THROUGH WITH ROW SEMANTICS,
+    Output_pivot_columns DESCRIPTOR,
+    Input_pivot_columns1 DESCRIPTOR,
+    Input_pivot_columns2 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns3 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns4 DESCRIPTOR DEFAULT NULL,
+    Input_pivot_columns5 DESCRIPTOR DEFAULT NULL
+)
+RETURNS TABLE
+DETERMINISTIC
+READS SQL DATA
+```
+
+But before we look at how you can invoke this, there's a few keywords here that
+are worth explaining...
+
+* `PASS THROUGH` means that the input data (and all of its rows) will be fully
+  available in the output. The alternative to this is `NO PASS THROUGH`.
+* `WITH ROW SEMANTICS` means that the result will be determined on a row-by-row
+  basis. The alternative to this is `WITH SET SEMANTICS`.
+
+And of course, the function takes some parameters, so a good function author
+defines what those parameters do.
+
+* 'Input' is the input table. It's any generic table.
+* 'Output_pivot_columns' is the names of the columns to be created in the pivot
+  table.
+* Input_pivot_columns are all the columns to be pivoted into the output columns.
+  The first parameter is required, but you can specify more groupings. The
+  number of input columns in a group to be pivoted and the number of output
+  columns must be the same.
+
+So you've got a PIVOT function, and you understand how to invoke it, so all you
+need to do is listen to [Ross from Friends](https://youtu.be/8w3wmQAMoxQ?t=82)
+and make it happen:
+
+```sql
+SELECT
+    D.id,
+    D.name,
+    P.accttype,
+    P.acctvalue
+FROM
+    TABLE(
+        Pivot(
+            Input_table => TABLE (My.Data) AS D,
+            Output_pivot_columns => DESCRIPTOR (accttype, acctvalue),
+            Input_pivot_columns1 => DESCRIPTOR (accttype1, acctvalue1),
+            Input_pivot_columns2 => DESCRIPTOR (accttype2, acctvalue2)
+        )
+    ) AS P;
+```
+
+If we presume we have this data in My.Data:
+
+| ID  | Name    | accttype1 | acctvalue1 | accttype2 | acctvalue2 |
+|-----|---------|-----------|------------|-----------|------------|
+| 123 | Alice   | external  | 20000      | internal  | 350        |
+| 234 | Bob     | external  | 25000      | internal  | 120        |
+
+The output of that query will be:
+
+| ID  | Name    | accttype | acctvalue |
+|-----|---------|----------|-----------|
+| 123 | Alice   | external | 20000     |
+| 123 | Alice   | internal | 350       |
+| 234 | Bob     | external | 25000     |
+| 234 | Bob     | internal | 120       |
+
+You can see the `PASS THROUGH` part of the signature working when you select
+D.id and D.name.
```

Comment:
With `PASS THROUGH`, all columns of My.Data should be available in the output. Here, the input pivot columns are missing. I noticed the same in the original example in the Technical Report.
I think we can remove the `PASS THROUGH` clause from this example, together with the explanation, and explain the `PASS THROUGH` / `NO PASS THROUGH` in the next example `ExecR`.

Reply:
I don't think they're missing? It's just that the query is only selecting a few specific columns, and isn't doing `SELECT *`

Reply:
Ah, right.